HopsanGenerator/src/generators/HopsanFMIGenerator.cpp

```diff
@@ -1842,72 +1713,110 @@ bool HopsanFMIGenerator::compileAndLinkFMU(const QString &fmuBuildPath, const QS
+    for(const QString includePath : mIncludePaths) {
+        makefileStream << QString(" -I\"%1\"").arg(includePath);
     }
-    linkBatchStream << " -o \""+outputLibraryFile+"\"\n";
-    linkBatchFile.close();
-
-    linkingOK = callProcess("cmd.exe", QStringList() << "/c" << "cd /d " + fmuBuildPath + " & link.bat");
-
+    makefileStream << "\n";
+    makefileStream << "CINCLUDES = -I\""+fmiLibDir+"/include\"\n";
+    makefileStream << "OUTPUT = \""+outputLibraryFile+"\"\n\n";
+    makefileStream << "CXXSRC = fmu_hopsan.cpp";
+    QStringList srcFiles = listHopsanCoreSourceFiles(fmuBuildPath) + listInternalLibrarySourceFiles(fmuBuildPath);
+    for(const QString& srcFile : srcFiles)
+    {
+        makefileStream << " " << srcFile;
+    }
+    makefileStream << "\n";
+    makefileStream << "CSRC = fmu"+vStr+"_model_cs.c\n";
+    makefileStream << "COBJ = fmu"+vStr+"_model_cs.o\n";
+    makefileStream << "\n\n";
+    makefileStream << "VPATH := $(sort  $(dir $(CXXSRC)))\n\n";
+    makefileStream << "CXXOBJ := $(patsubst %.cpp, %.o, $(notdir $(CXXSRC)))\n\n";
+    makefileStream << "all: 	$(CXXOBJ)\n";
+    makefileStream << "\t$(CC) $(CFLAGS) $(CINCLUDES) -c $(CSRC) -o $(COBJ)\n";
+    makefileStream << "\t$(CXX) $(LFLAGS) $(COBJ) $(CXXOBJ) -o $(OUTPUT)\n\n";
+    makefileStream << "%.o : %.cpp Makefile\n";
+    makefileStream << "\t$(CXX) $(CXXFLAGS) $(CXXINCLUDES) -c $< -o $@\n\n";
+    makefile.close();
+
+    printMessage("Generating compilation script");
+
+    QFile compileScriptFile;
+    QString scriptExt;
+#ifdef _WIN32
+    scriptExt = "ps1";
 #else
-    QFile linkBatchFile;
-    linkBatchFile.setFileName(fmuBuildPath + "/link.sh");
-    if(!linkBatchFile.open(QIODevice::WriteOnly | QIODevice::Text))
+    scriptExt = "sh";
+#endif
+    compileScriptFile.setFileName(fmuBuildPath + "/compile."+scriptExt);
+    if(!compileScriptFile.open(QIODevice::WriteOnly | QIODevice::Text))
     {
-        printErrorMessage("Failed to open link.sh for writing.");
+        printErrorMessage("Failed to open compile."+scriptExt+" for writing.");
         return false;
     }
+
     //Write the compilation script file
-    QTextStream linkBatchStream(&linkBatchFile);
-    linkBatchStream << mCompilerSelection.path+"g++ -fPIC -w -shared -static-libgcc -Wl,--rpath,'$ORIGIN/.' " << "fmu"+vStr+"_model_cs.o";
-    Q_FOREACH(const QString &objFile, objectFiles)
-    {
-        linkBatchStream << " " << objFile;
-    }
-    for(const QString linkPaths : mLinkPaths) {
-        linkBatchStream << " -L\"" << linkPaths << "\"";
-    }
-    for(const QString linkLibrary : mLinkLibraries) {
-        linkBatchStream << " -l" << linkLibrary;
-    }
-    linkBatchStream << " -o \""+outputLibraryFile+"\"\n";
-    linkBatchFile.close();
+    QTextStream compileScriptStream(&compileScriptFile);
+#ifdef _WIN32
+    compileScriptStream << "$env:Path = \""+mCompilerSelection.path+";$env:Path\"\n";
+    compileScriptStream << "mingw32-make -j all\n";
```

Comment:
Maybe should add a limit to the number of threads, maybe 16 is reasonable nowdays.
Hopsan is limited in size, but if you are low on ram you may still run out, since it will build all .cpp files in parallel.
(Both for windows and linux)

Reply:
I didn't want to hard-core this limit, but perhaps it is needed for memory reasons. I read somewhere that the optimal value is roughtly 1.5 x number of processors to achieve good load balancing. But we can go with 16 jobs for now I guess.